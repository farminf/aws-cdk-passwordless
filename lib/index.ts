import * as path from "path";
import cdk = require("@aws-cdk/core");
import cognito = require("@aws-cdk/aws-cognito");
import iam = require("@aws-cdk/aws-iam");
import lambda = require("@aws-cdk/aws-lambda");

export interface CdkPasswordlessProps {
  userPoolClientName?: string;
  verifiedDomains?: Array<string>;
  mailSubject?: string;
}

export class CdkPasswordless extends cdk.Construct {
  /** @returns the ARN of the SQS queue */
  public readonly userPool: cognito.UserPool;
  public readonly userPoolClient: cognito.UserPoolClient;

  constructor(
    scope: cdk.Construct,
    id: string,
    props: CdkPasswordlessProps = {}
  ) {
    super(scope, id);

    const { userPoolClientName, verifiedDomains, mailSubject } = props;

    const lambdaRole = new iam.Role(this, "lambdaRole", {
      assumedBy: new iam.CompositePrincipal(
        new iam.ServicePrincipal("lambda.amazonaws.com"),
        new iam.ServicePrincipal("cognito-idp.amazonaws.com")
      ),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AWSLambdaBasicExecutionRole"
        )
      ]
    });

    const cognitoEventsLambda = new lambda.Function(
      this,
      "cognitoEventsLambda",
      {
        code: lambda.Code.asset(path.resolve(__dirname, "functions")),
        description:
          "This function auto-confirms users and their email addresses during sign-up",
        handler: "cognitoEvents.handler",
        runtime: lambda.Runtime.NODEJS_10_X,
        role: lambdaRole,
        environment: {
          VERIFIED_DOMAINS: JSON.stringify(verifiedDomains) || undefined,
          SIGNINSUBJECT: mailSubject || undefined
        }
      }
    );

    const userPool = new cognito.UserPool(this, "userPool", {
      lambdaTriggers: {
        preSignUp: cognitoEventsLambda
      },
      autoVerifiedAttributes: [cognito.UserPoolAttribute.EMAIL],
      signInType: cognito.SignInType.EMAIL
    });

    const userPoolClient = new cognito.UserPoolClient(this, "userPoolClient", {
      userPoolClientName: userPoolClientName || "passwordless",
      generateSecret: false,
      userPool: userPool
    });

    this.userPool = userPool;
    this.userPoolClient = userPoolClient;

    new cdk.CfnOutput(this, "userPoolIdOutput", {
      exportName: "passwordLessUserPoolId",
      value: this.userPool.userPoolId
    });
    new cdk.CfnOutput(this, "userPoolArnOutput", {
      exportName: "passwordLessUserPoolArn",
      value: this.userPool.userPoolArn
    });
    new cdk.CfnOutput(this, "userPoolClientIdOutput", {
      exportName: "passwordLessUserPoolClientId",
      value: this.userPoolClient.userPoolClientId
    });
  }
}
