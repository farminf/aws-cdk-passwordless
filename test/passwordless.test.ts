import { Stack, App } from "@aws-cdk/core";
import { CdkPasswordless } from "../lib/index";

test("testing passwordless resources", () => {
  const app = new App();
  const stack = new Stack(app, "test");
  const myPasswordlessTest = new CdkPasswordless(
    stack,
    "myPasswordlessTest",
    {}
  );

  expect(app.synth().getStack("test").template).toMatchSnapshot();
});
