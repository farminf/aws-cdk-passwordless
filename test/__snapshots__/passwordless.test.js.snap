// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing passwordless resources 1`] = `
Object {
  "Outputs": Object {
    "myPasswordlessTestuserPoolArnOutput04B8D94D": Object {
      "Export": Object {
        "Name": "passwordLessUserPoolArn",
      },
      "Value": Object {
        "Fn::GetAtt": Array [
          "myPasswordlessTestuserPool18816AD8",
          "Arn",
        ],
      },
    },
    "myPasswordlessTestuserPoolClientIdOutput76E5E879": Object {
      "Export": Object {
        "Name": "passwordLessUserPoolClientId",
      },
      "Value": Object {
        "Ref": "myPasswordlessTestuserPoolClient442FD4CD",
      },
    },
    "myPasswordlessTestuserPoolIdOutputD2D99BBE": Object {
      "Export": Object {
        "Name": "passwordLessUserPoolId",
      },
      "Value": Object {
        "Ref": "myPasswordlessTestuserPool18816AD8",
      },
    },
  },
  "Parameters": Object {
    "myPasswordlessTestcognitoEventsLambdaCodeArtifactHash8198C4D7": Object {
      "Description": "Artifact hash for asset \\"test/myPasswordlessTest/cognitoEventsLambda/Code\\"",
      "Type": "String",
    },
    "myPasswordlessTestcognitoEventsLambdaCodeS3Bucket4E0CA048": Object {
      "Description": "S3 bucket for asset \\"test/myPasswordlessTest/cognitoEventsLambda/Code\\"",
      "Type": "String",
    },
    "myPasswordlessTestcognitoEventsLambdaCodeS3VersionKeyB7C1D3F9": Object {
      "Description": "S3 key for asset version \\"test/myPasswordlessTest/cognitoEventsLambda/Code\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "myPasswordlessTestcognitoEventsLambdaC85C55A5": Object {
      "DependsOn": Array [
        "myPasswordlessTestlambdaRole51C60D45",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "myPasswordlessTestcognitoEventsLambdaCodeS3Bucket4E0CA048",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "myPasswordlessTestcognitoEventsLambdaCodeS3VersionKeyB7C1D3F9",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "myPasswordlessTestcognitoEventsLambdaCodeS3VersionKeyB7C1D3F9",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "This function auto-confirms users and their email addresses during sign-up",
        "Environment": Object {
          "Variables": Object {},
        },
        "Handler": "cognitoEvents.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "myPasswordlessTestlambdaRole51C60D45",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "myPasswordlessTestcognitoEventsLambdaCustomMessageCognito774653EB": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "myPasswordlessTestcognitoEventsLambdaC85C55A5",
            "Arn",
          ],
        },
        "Principal": "cognito-idp.amazonaws.com",
      },
      "Type": "AWS::Lambda::Permission",
    },
    "myPasswordlessTestcognitoEventsLambdaPreSignUpCognito67096823": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "myPasswordlessTestcognitoEventsLambdaC85C55A5",
            "Arn",
          ],
        },
        "Principal": "cognito-idp.amazonaws.com",
      },
      "Type": "AWS::Lambda::Permission",
    },
    "myPasswordlessTestlambdaRole51C60D45": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Array [
                  Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "lambda.",
                        Object {
                          "Ref": "AWS::URLSuffix",
                        },
                      ],
                    ],
                  },
                  Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "cognito-idp.",
                        Object {
                          "Ref": "AWS::URLSuffix",
                        },
                      ],
                    ],
                  },
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "myPasswordlessTestuserPool18816AD8": Object {
      "Properties": Object {
        "AutoVerifiedAttributes": Array [
          "email",
        ],
        "LambdaConfig": Object {
          "CustomMessage": Object {
            "Fn::GetAtt": Array [
              "myPasswordlessTestcognitoEventsLambdaC85C55A5",
              "Arn",
            ],
          },
          "PreSignUp": Object {
            "Fn::GetAtt": Array [
              "myPasswordlessTestcognitoEventsLambdaC85C55A5",
              "Arn",
            ],
          },
        },
        "UsernameAttributes": Array [
          "email",
        ],
      },
      "Type": "AWS::Cognito::UserPool",
    },
    "myPasswordlessTestuserPoolClient442FD4CD": Object {
      "Properties": Object {
        "ClientName": "passwordless",
        "GenerateSecret": false,
        "UserPoolId": Object {
          "Ref": "myPasswordlessTestuserPool18816AD8",
        },
      },
      "Type": "AWS::Cognito::UserPoolClient",
    },
  },
}
`;
